Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1465 @@
+  @Equiv.decidableEq (SkewMonoidAlgebra k G) _ toFinsuppAddEquiv.toEquiv Finsupp.instDecidableEq
+
+/-- `sum f g` is the sum of `g a (f a)` over the support of `f`. -/
+def sum {N : Type*} [AddCommMonoid N] (f : SkewMonoidAlgebra k G) (g : G → k → N) : N :=
+  f.toFinsupp.sum g
+
+theorem sum_def {N : Type*} [AddCommMonoid N] (f : SkewMonoidAlgebra k G) (g : G → k → N) :
+    sum f g = Finsupp.sum f.toFinsupp g := rfl
+
+@[simp]
+theorem sum_single_index {N} [AddCommMonoid N] {a : G} {b : k} {h : G → k → N}
+    (h_zero : h a 0 = 0) : (SkewMonoidAlgebra.single a b).sum h = h a b :=
+  Finsupp.sum_single_index h_zero
+
+theorem map_sum' {N P : Type*} [AddCommMonoid N] [AddCommMonoid P] {H : Type*} [FunLike H N P]
+    [AddMonoidHomClass H N P] (h : H) (f : SkewMonoidAlgebra k G) (g : G → k → N) :
+    h (sum f g) = sum f fun (a : G) (b : k) => h (g a b) :=
+  map_sum h _ _
+
+theorem ofFinsupp_sum {ι : Type*} (s : Finset ι) (f : ι → (G →₀ k)) :
+    (⟨Finset.sum s f⟩ : SkewMonoidAlgebra k G) = Finset.sum s (fun i ↦ ⟨f i⟩) :=
+  map_sum toFinsuppAddEquiv.symm f s
+
+@[simp]
+theorem toFinsupp_sum {ι : Type*} (s : Finset ι) (f : ι → SkewMonoidAlgebra k G) :
+    (Finset.sum s f : SkewMonoidAlgebra k G).toFinsupp = Finset.sum s (fun i ↦ (f i).toFinsupp) :=
+  map_sum toFinsuppAddEquiv f s
+
+theorem toFinsupp_sum' {k' G' : Type*} [AddCommMonoid k'] (f : SkewMonoidAlgebra k' G)
+    (g : G → k' → SkewMonoidAlgebra k G') :
+    (sum f g).toFinsupp = Finsupp.sum f.toFinsupp (fun (a : G) (b : k') ↦ (g a b).toFinsupp) :=
+  map_sum toFinsuppAddEquiv (fun (a : G) => g a (f a)) f.toFinsupp.support
+
+theorem sum_single (f : SkewMonoidAlgebra k G) : f.sum single = f :=
+  (ofFinsupp_sum _ _).symm.trans (congr_arg _ <| Finsupp.sum_single _)
+
+theorem sum_add_index' {S : Type*} [AddCommMonoid S] {f g : SkewMonoidAlgebra k G} {h : G → k → S}
+    (hf : ∀ i, h i 0 = 0) (h_add : ∀ a b₁ b₂, h a (b₁ + b₂) = h a b₁ + h a b₂) :
+    (f + g).sum h = f.sum h + g.sum h := by
+  rw [show f + g = ⟨f.toFinsupp + g.toFinsupp⟩ from add_def f g]
+  exact Finsupp.sum_add_index' hf h_add
+
+theorem sum_add_index {S : Type*} [DecidableEq G] [AddCommMonoid S]
+    {f g : SkewMonoidAlgebra k G} {h : G → k → S} (h_zero : ∀ a ∈ f.support ∪ g.support, h a 0 = 0)
+    (h_add : ∀ a ∈ f.support ∪ g.support, ∀ (b₁ b₂ : k), h a (b₁ + b₂) = h a b₁ + h a b₂) :
+    (f + g).sum h = f.sum h + g.sum h := by
+  rw [show f + g = ⟨f.toFinsupp + g.toFinsupp⟩ from add_def f g]
+  exact Finsupp.sum_add_index h_zero h_add
+
+@[simp]
+theorem sum_add' {S : Type*} [AddCommMonoid S] (p : SkewMonoidAlgebra k G) (f g : G → k → S) :
+    p.sum (f + g) = p.sum f + p.sum g := Finsupp.sum_add
+
+@[simp]
+theorem sum_add {S : Type*} [AddCommMonoid S] (p : SkewMonoidAlgebra k G) (f g : G → k → S) :
+    (p.sum fun n x => f n x + g n x) = p.sum f + p.sum g :=
+  sum_add' _ _ _
+
+theorem sum_smul_index' {N : Type*} [DistribSMul R k] [AddCommMonoid N]
```

Comment:
why the ' in the name?

Reply:
This comes from `Finsupp` and `Polynomial`. It is thought to be useful to have the two forms of the lemma. So we have `sum_smul_index` and `sum_smul_index'`